pkgs/by-name/lo/localai/package.nix

```diff
@@ -0,0 +1,245 @@
+    owner = "mudler";
+    repo = "go-stable-diffusion";
+    rev = "d5d2be8e7e395c2d73ceef61e6fe8d240f2cd831";
+    hash = "sha256-MbVYeWQF/aJNsg2NpTMVx5tD31BK5pQ8Zg92uoWRkcU=";
+    fetchSubmodules = true;
+  };
+
+  go-tiny-dream = fetchFromGitHub {
+    owner = "M0Rf30";
+    repo = "go-tiny-dream";
+    rev = "772a9c0d9aaf768290e63cca3c904fe69faf677a";
+    hash = "sha256-r+wzFIjaI6cxAm/eXN3q8LRZZz+lE5EA4lCTk5+ZnIY=";
+    fetchSubmodules = true;
+  };
+
+  go-tiny-dream' = runCommand "go_tiny_dream_src" { } ''
+    cp -r --no-preserve=mode,ownership ${go-tiny-dream} $out
+    sed -i $out/Makefile \
+      -e 's;lib/libncnn;lib64/libncnn;g'
+  '';
+
+  GO_TAGS = lib.optional enableTinydream "tinydream"
+    ++ lib.optional enableTts "tts"
+    ++ lib.optional enableStablediffusion "stablediffusion";
+in
+(buildGoModule.override { stdenv = gcc13Stdenv; }) rec {
+  pname = "localai";
+  version = "2.8.0";
+
+  src = fetchFromGitHub {
+    owner = "go-skynet";
+    repo = "LocalAI";
+    rev = "v${version}";
+    hash = "sha256-JwJX/zpZUKC7arntGe/quN8YPSH2C4UL9KM7YhOFTGA=";
+  };
+
+  patches = lib.optional (version == "2.8.0")
+    (fetchpatch {
+      name = "fix-drop-unused-code";
+      url = "https://github.com/mudler/LocalAI/pull/1697/commits/a4d4c26ee821482cac89fefa511808e987434654.patch";
+      hash = "sha256-37qURvdTI6lFkAepf7qsxhOVa1CXxIGSieorHxLSgRE=";
+    });
+
+  vendorHash = "sha256-XmouAnK+N41SgdUdY1RUHrG5DKLlAZNpimQCvL50zKA=";
+
+  # Workaround for
+  # `cc1plus: error: '-Wformat-security' ignored without '-Wformat' [-Werror=format-security]`
+  # when building jtreg
+  env.NIX_CFLAGS_COMPILE = "-Wformat"
+    + lib.optionalString enableStablediffusion " -isystem ${opencv}/include/opencv4";
+
+  postPatch =
+    let
+      cp = "cp -r --no-preserve=mode,ownership";
+    in
+    ''
+      sed -i Makefile \
+        -e 's;git clone.*go-llama$;${cp} ${go-llama} sources/go-llama;' \
+        -e 's;git clone.*go-llama-ggml$;${cp} ${go-llama-ggml} sources/go-llama-ggml;' \
+        -e 's;git clone.*gpt4all$;${cp} ${gpt4all} sources/gpt4all;' \
+        -e 's;git clone.*go-piper$;${cp} ${go-piper} sources/go-piper;' \
+        -e 's;git clone.*go-rwkv$;${cp} ${go-rwkv} sources/go-rwkv;' \
+        -e 's;git clone.*whisper\.cpp$;${cp} ${whisper} sources/whisper\.cpp;' \
+        -e 's;git clone.*go-bert$;${cp} ${go-bert} sources/go-bert;' \
+        -e 's;git clone.*diffusion$;${cp} ${go-stable-diffusion} sources/go-stable-diffusion;' \
+        -e 's;git clone.*go-tiny-dream$;${cp} ${go-tiny-dream'} sources/go-tiny-dream;' \
+        -e 's, && git checkout.*,,g' \
+        -e '/mod download/ d' \
+
+      sed -i backend/cpp/llama/Makefile \
+        -e 's;git clone.*llama\.cpp$;${cp} ${llama_cpp'} llama\.cpp;' \
+        -e 's, && git checkout.*,,g' \
+
+    ''
+  ;
+
+  modBuildPhase = ''
+    mkdir sources
+    make prepare-sources
+    go mod tidy -v
+  '';
+
+  proxyVendor = true;
+
+  buildPhase = ''
+    mkdir sources
```

Comment:
you should call the hooks, e.g. `runHook preBuild`

Reply:
can this be done via `makeFlags` and the inherited buildPhase ? doesn't upstream run `mkdir sources` ?

Reply:
The `buildGoModule` derivation does not call `make`, therefor we have to override `buildPhase`.

The `makeFlags` could be separated, but then the custom `buildPhase` has to respect the environment variable.

Regarding directory sources, the original `Makefile` calls `git clone`, which creates the sources directory. The patched `Makefile` calls `cp`, which does not create the leading directories.